src/agent/src/policy.rs

```diff
@@ -0,0 +1,218 @@
+            }
+
+            // Set-up the default policy.
+            if let Ok(_) = self
+                .opa_client
+                .put(&self.policy_path)
+                .body(policy.clone())
+                .send()
+                .await
+            {
+                // Check if requests causing policy errors should actually
+                // be allowed. That is an insecure configuration but is
+                // useful for allowing insecure pods to start, then connect to
+                // them and inspect Guest logs for the root cause of a failure.
+                if let Ok(allow_failures) = self
+                    .post_query("AllowRequestsFailingPolicy", EMPTY_JSON_INPUT)
+                    .await
+                {
+                    self.allow_failures = allow_failures;
+                }
+                return Ok(());
+            }
+        }
+        Err(anyhow!("failed to connect to OPA"))
+    }
+
+    /// Ask OPA to check if an API call should be allowed or not.
+    pub async fn is_allowed_endpoint(&mut self, ep: &str, request: &str) -> bool {
+        let post_input = "{\"input\":".to_string() + request + "}";
+
+        #[cfg(feature = "agent-policy-debug")]
+        self.log_opa_input(ep, &post_input).await;
+
+        self.post_query(ep, &post_input).await.unwrap_or(false)
+    }
+
+    /// Replace the Policy in OPA.
+    pub async fn set_policy(&mut self, policy: &str) -> Result<()> {
+        // Delete the old rules.
+        self.opa_client
+            .delete(&self.policy_path)
+            .send()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        // Put the new rules.
+        self.opa_client
+            .put(&self.policy_path)
+            .body(policy.to_string())
+            .send()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        // Check if requests causing policy errors should actually be allowed.
+        // That is an insecure configuration but is useful for allowing insecure
+        // pods to start, then connect to them and inspect Guest logs for the
+        // root cause of a failure.
+        self.allow_failures = self
+            .post_query("AllowRequestsFailingPolicy", EMPTY_JSON_INPUT)
+            .await?;
+        Ok(())
+    }
+
+    // Post query to OPA.
+    async fn post_query(&mut self, ep: &str, post_input: &str) -> Result<bool> {
+        debug!(sl!(), "policy check: {}", ep);
+
+        let uri = self.query_path.clone() + ep;
+        let response = self
+            .opa_client
+            .post(uri)
+            .body(post_input.to_string())
+            .send()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        if response.status() != http::StatusCode::OK {
+            return Err(anyhow!(
+                "policy: post_query: POST response status {}",
+                response.status()
+            ));
+        }
+
+        let http_response = response.text().await.unwrap();
```

Comment:
should we raise this error to upper caller?

Reply:
I think unwrap is correct - but I don't mind sending this error to the caller if other folks prefer it that way.

Reply:
I replaced the unwrap calls, but I added an abort call - please see https://github.com/kata-containers/kata-containers/pull/7574/commits/5f1bbe4a79d3d6d07ec25e0253c23b3f6dc758b0